[TGUI] Pacman

RedWolfie opened the pull request:
## What Does This PR Do
Replaces NanoUI of the pacman generators with TGUI 

## Why It's Good For The Game
Gets rid of Nano

## Images of changes
On
![pacman-demo9](https://user-images.githubusercontent.com/13683223/93888761-b6efc300-fce8-11ea-9708-d944e5cf4782.png)

Off
![pacman-demo8](https://user-images.githubusercontent.com/13683223/93888762-b7885980-fce8-11ea-8572-8768f2ad7a13.png)

Errors
![pacman-demo11](https://user-images.githubusercontent.com/13683223/93888756-b6572c80-fce8-11ea-9d80-cbc45b5db3c8.png)
![pacman-demo12](https://user-images.githubusercontent.com/13683223/93888955-edc5d900-fce8-11ea-9cbc-58205f31b6be.png)


Old Nano for reference:
![pacman-ref](https://user-images.githubusercontent.com/13683223/93649560-8e2bad00-fa0c-11ea-8bef-ac53b58cb560.png)


## Changelog
:cl:
add: Added TGUI for Pacman generators
/:cl:

TDSSS commented:
Why are we measuring plasma in liters now?

Also, the malfunction Ui seems a bit empty.

Kyep commented on tgui/packages/tgui/interfaces/Pacman.js:
```diff
@@ -0,0 +1,174 @@
+  } = data;
+  let fuelRatio = fuel_stored / fuel_cap;
+  let tmpRatio = tmp_current / tmp_max; // temperature ratio
+  let output_watts = output_set * power_gen;
```
```suggestion
  let output_watts = (output_set * power_gen).toString().replace(/(\d)(?=(\d{3})+(?!\d))/g, '$1,');;
```
Either use something like this to add a thousands separator, or just show the value in kW.
Otherwise you get output like "200000 W".

AffectedArc07 replied:
Take a look at `tgui\packages\tgui\interfaces\Smes.js` and make usre of `formatPower()` please

RedWolfie replied:
done

Kyep commented on tgui/packages/tgui/interfaces/Pacman.js:
```diff
@@ -0,0 +1,174 @@
+                    <LabeledList.Item label="Fuel depletion">
+                      {(fuel_usage
+                        ? usage
+                        : "Infinite")}
```
```suggestion
                        : "N/A")}
```

RedWolfie replied:
done

Kyep commented on code/modules/power/port_gen.dm:
```diff
@@ -312,78 +312,72 @@
-	var/data[0]
+/obj/machinery/power/port_gen/pacman/tgui_data(mob/user)
+	var/list/data = list()
+	var/sheetVolume = 1000 //cm3
```
No global or define for this?

RedWolfie replied:
added a define

Kyep commented on code/modules/power/port_gen.dm:
```diff
@@ -312,78 +312,72 @@
+	switch(action)
+		if("toggle_power")
+			if(!HasFuel())
+				atom_say("Out of fuel. ")
```
Space at the end.
Also this should be unreachable - "on" button should be disabled (and an 'no fuel' shown in interface) if generator has no fuel.

RedWolfie replied:
done

Kyep commented on code/modules/power/port_gen.dm:
```diff
@@ -312,78 +312,72 @@
+		if("eject_fuel")
+			DropFuel()
+		if("change_power")
+			var/newPower = text2num(params["change_power"])
```
No validation on this input?

RedWolfie replied:
clamp added

AffectedArc07 commented:
Any reason for the control duplication here?
![image](https://user-images.githubusercontent.com/25063394/93664025-b1ce1200-fa63-11ea-85ae-e5772a0259de.png)

AffectedArc07 commented on code/modules/power/port_gen.dm:
```diff
@@ -312,78 +312,72 @@
+		if("eject_fuel")
+			DropFuel()
+		if("change_power")
+			var/newPower = text2num(params["change_power"])
```
This **needs** wrapping inside of a `clamp()`.

RedWolfie replied:
added

AffectedArc07 commented on tgui/packages/tgui/interfaces/Pacman.js:
```diff
@@ -0,0 +1,174 @@
+          <Section title="Status">
+            {!!broken && (
+              <Box color="orange">
+                The generator is malfunctioning.
```
```suggestion
                The generator is malfunctioning!
```
Its a bad thing, so I feel an `!` is warranted

RedWolfie replied:
Done

AffectedArc07 commented on tgui/packages/tgui/interfaces/Pacman.js:
```diff
@@ -0,0 +1,174 @@
+                    </LabeledList.Item>
+                    <LabeledList.Item label="Status">
+                      {tmp_overheat > 50 && (
+                        <Box color="red"> CRITICAL OVERHEAT! </Box>
+                      )}
+                      {tmp_overheat > 20 && tmp_overheat <= 50 && (
+                        <Box color="orange"> WARNING: Overheating! </Box>
+                      )}
+                      {tmp_overheat > 1 && tmp_overheat <= 20 && (
+                        <Box color="orange"> Temperature High </Box>
+                      )}
+                      {tmp_overheat === 0 && (
+                        <Box color="green"> Optimal </Box>
```
Any reason for the spacing before and after the `<Box>` tags

RedWolfie replied:
No, removed

RedWolfie commented:
> Any reason for the control duplication here?
> ![image](https://user-images.githubusercontent.com/25063394/93664025-b1ce1200-fa63-11ea-85ae-e5772a0259de.png)

Yes, I thought the knob looked nice, but it has bad value display and bad usability, so I added the number input to compensate for those shortcomings. 

> Why are we measuring plasma in liters now?
> 
> Also, the malfunction Ui seems a bit empty.

Changed from liters to cubic dm
![pacman-demo6](https://user-images.githubusercontent.com/13683223/93668315-eeadff00-fa8b-11ea-854e-0f38a057a2f3.png)

AffectedArc07 commented:
Duplicate controls for the same thing are bad. You can give the knob a tooltip on what its currently set at if you want to keep it, otherwise, choose one or the other

AffectedArc07 commented:
Please update your screenshots
